extension/src/runners/runnerExecutor.ts

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+import * as cp from 'child_process';
+import { ExtensionContext, window } from 'vscode';
+import { testCodeLensProvider } from '../codeLensProvider';
+import { CHILD_PROCESS_MAX_BUFFER_SIZE } from '../constants/configs';
+import { ITestItem, TestKind } from '../protocols';
+import { IExecutionConfig } from '../runConfigs';
+import { testResultManager } from '../testResultManager';
+import { testStatusBarProvider } from '../testStatusBarProvider';
+import { killProcess } from '../utils/cpUtils';
+import { ITestRunner } from './ITestRunner';
+import { JUnit4Runner } from './junit4Runner/Junit4Runner';
+import { JUnit5Runner } from './junit5Runner/JUnit5Runner';
+import { ITestResult } from './models';
+import { TestNGRunner } from './testngRunner/TestNGRunner';
+
+export class RunnerExecutor {
+    private isRunning: boolean;
+    private preLaunchTask: cp.ChildProcess | undefined;
+    private runnerMap: Map<ITestRunner, ITestItem[]> | undefined;
+
+    constructor(private _javaHome: string, private _context: ExtensionContext) {
+    }
+
+    public async run(testItems: ITestItem[], isDebug: boolean = false, config?: IExecutionConfig): Promise<void> {
+        if (this.isRunning) {
+            window.showInformationMessage('A test session is currently running. Please wait until it finishes.');
+            return;
+        }
+        this.isRunning = true;
+        testStatusBarProvider.showRunningTest();
+        try {
+            this.runnerMap = this.classifyTestsByKind(testItems);
+            for (const [runner, tests] of this.runnerMap.entries()) {
+                if (config && config.preLaunchTask.length > 0) {
+                    this.preLaunchTask = cp.exec(
+                        config.preLaunchTask,
+                        {
+                            maxBuffer: CHILD_PROCESS_MAX_BUFFER_SIZE,
+                            cwd: config.workingDirectory,
+                        },
+                    );
+                    await this.execPreLaunchTask();
+                }
+                await runner.setup(tests, isDebug, config);
+                const results: ITestResult[] = await runner.run();
+                testResultManager.storeResult(...results);
+                testStatusBarProvider.showTestResult(results);
+                testCodeLensProvider.refresh();
+            }
+        } catch (error) {
+            window.showErrorMessage(`${error}`);
+        } finally {
+            await this.cleanUp();
+        }
+    }
+
+    private async cleanUp(): Promise<void> {
+        try {
+            if (this.preLaunchTask) {
+                await killProcess(this.preLaunchTask);
+                this.preLaunchTask = undefined;
+            }
+
```

Comment:
Should swallow here? No matter what happens insider the preLaunchTask, need clean up the resource anyway.